src/environments/array.js

```diff
@@ -0,0 +1,475 @@
+// @flow
+import buildCommon, {makeSpan} from "../buildCommon";
+import defineEnvironment from "../defineEnvironment";
+import mathMLTree from "../mathMLTree";
+import ParseError from "../ParseError";
+import ParseNode from "../ParseNode";
+import {calculateSize} from "../units";
+import utils from "../utils";
+
+import * as html from "../buildHTML";
+import * as mml from "../buildMathML";
+
+import type Parser from "../Parser";
+import type {StyleStr} from "../types";
+
+// Data stored in the ParseNode associated with the environment.
+type AlignSpec = { type: "separator", separator: string } | {
+    type: "align",
+    align: string,
+    pregap?: number,
+    postgap?: number,
+};
+type ArrayEnvNodeData = {
+    type: "array",
+    hskipBeforeAndAfter?: boolean,
+    arraystretch?: number,
+    addJot?: boolean,
+    cols?: AlignSpec[],
+    // These fields are always set, but not on struct construction
+    // initialization.
+    body?: ParseNode[][], // List of rows in the (2D) array.
+    rowGaps?: number[],
+};
+
+/**
+ * Parse the body of the environment, with rows delimited by \\ and
+ * columns delimited by &, and create a nested list in row-major order
+ * with one group per cell.  If given an optional argument style
+ * ("text", "display", etc.), then each cell is cast into that style.
+ */
+function parseArray(
+    parser: Parser,
+    result: ArrayEnvNodeData,
+    style: StyleStr,
+): ParseNode {
+    let row = [];
+    const body = [row];
+    const rowGaps = [];
+    for (;;) {
```

Comment:
I know this is a copypasta, but `while(true)` would be clearer here.

Reply:
It was originally `while(true)`, but a previous commit changed it after removing a top level eslint-disable.  I can add it back in with a eslint-disable-line.